File: hdf-generators/go/integration_test.go
package generators

import (
	"encoding/json"
	"os"
	"strings"
	"testing"

	hdf "github.com/mitre/hdf-schema"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func loadFixture(t *testing.T) hdf.HDFBaseline {
	t.Helper()
	data, err := os.ReadFile("testdata/win2022-stig-baseline.json")
	require.NoError(t, err, "fixture file must exist (symlink from test/fixtures/)")
	var baseline hdf.HDFBaseline
	require.NoError(t, json.Unmarshal(data, &baseline))
	return baseline
}

func TestIntegration_LoadFixture(t *testing.T) {
	baseline := loadFixture(t)
	assert.Equal(t, "microsoft-windows-server-2022-stig-baseline", baseline.Name)
	assert.NotNil(t, baseline.Title)
	assert.Equal(t, "Microsoft Windows Server 2022 Security Technical Implementation Guide", *baseline.Title)
	assert.Len(t, baseline.Requirements, 18)
}

func TestIntegration_GeneratesAllControls(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)
	assert.Len(t, profile.Controls, 18)
	for _, req := range baseline.Requirements {
		filename := "controls/" + req.ID + ".rb"
		_, ok := profile.Controls[filename]
		assert.True(t, ok, "missing control file for %s", req.ID)
	}
}

func TestIntegration_ValidRubyBlocks(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)
	for filename, ruby := range profile.Controls {
		id := strings.TrimSuffix(strings.TrimPrefix(filename, "controls/"), ".rb")
		assert.True(t, strings.HasPrefix(ruby, "control '"+id+"' do\n"),
			"control %s should start with control block", id)
		assert.True(t, strings.HasSuffix(ruby, "end\n"),
			"control %s should end with 'end'", id)
	}
}

func TestIntegration_STIGMetadataInTags(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)

	sv238 := profile.Controls["controls/SV-254238.rb"]
	assert.Contains(t, sv238, "tag cci: ['CCI-000366']")
	assert.Contains(t, sv238, "tag nist: ['CM-6 b']")
	assert.Contains(t, sv238, "tag severity: 'medium'")
	assert.Contains(t, sv238, "tag stig_id: 'WN22-00-000010'")
}

func TestIntegration_MultipleCCIsAndNIST(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)

	sv240 := profile.Controls["controls/SV-254240.rb"]
	assert.Contains(t, sv240, "'CCI-000366'")
	assert.Contains(t, sv240, "'CCI-001312'")
	assert.Contains(t, sv240, "'CM-6 b'")
	assert.Contains(t, sv240, "'SI-11 a'")
}

func TestIntegration_CheckAndFixDescriptions(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)

	sv238 := profile.Controls["controls/SV-254238.rb"]
	assert.Contains(t, sv238, "desc 'check'")
	assert.Contains(t, sv238, "desc 'fix'")
}

func TestIntegration_InSpecYml(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)

	assert.Contains(t, profile.InSpecYml, "name: microsoft-windows-server-2022-stig-baseline")
	assert.Contains(t, profile.InSpecYml, "title: Microsoft Windows Server 2022 Security Technical Implementation Guide")
	assert.Contains(t, profile.InSpecYml, "version: '2.7.0'")
	assert.Contains(t, profile.InSpecYml, "maintainer: MITRE SAF Team")
	assert.Contains(t, profile.InSpecYml, "license: Apache-2.0")
}

func TestIntegration_SingleFile(t *testing.T) {
	baseline := loadFixture(t)
	opts := &GeneratorOptions{SingleFile: true}
	profile := GenerateInSpecProfile(baseline, opts)

	assert.Len(t, profile.Controls, 1)
	content, ok := profile.Controls["controls/controls.rb"]
	assert.True(t, ok)
	for _, req := range baseline.Requirements {
		assert.Contains(t, content, "control '"+req.ID+"' do")
	}
}

func TestIntegration_MetadataOverrides(t *testing.T) {
	baseline := loadFixture(t)
	opts := &GeneratorOptions{
		Metadata: &ProfileMetadata{
			Maintainer: "Custom Team",
			Version:    "99.0.0",
		},
	}
	profile := GenerateInSpecProfile(baseline, opts)

	assert.Contains(t, profile.InSpecYml, "maintainer: Custom Team")
	assert.Contains(t, profile.InSpecYml, "version: '99.0.0'")
	assert.NotContains(t, profile.InSpecYml, "maintainer: MITRE SAF Team")
	assert.NotContains(t, profile.InSpecYml, "version: '2.7.0'")
}

func TestIntegration_HighImpactControls(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)

	sv240 := profile.Controls["controls/SV-254240.rb"]
	assert.Contains(t, sv240, "impact 0.7")
	assert.Contains(t, sv240, "tag severity: 'high'")
}

func TestIntegration_MultiLineDescriptions(t *testing.T) {
	baseline := loadFixture(t)
	profile := GenerateInSpecProfile(baseline, nil)

	sv240 := profile.Controls["controls/SV-254240.rb"]
	assert.Contains(t, sv240, "web browser")
	assert.Contains(t, sv240, "administrative account")
}

File: hdf-generators/test/integration.test.ts
import { readFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { describe, it, expect } from 'vitest';
import type { HdfBaseline } from '@mitre/hdf-schema';
import { generateInSpecProfile } from '../src/profile-generator.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const fixturePath = resolve(__dirname, 'fixtures', 'win2022-stig-baseline.json');

function loadFixture(): HdfBaseline {
  return JSON.parse(readFileSync(fixturePath, 'utf-8')) as HdfBaseline;
}

describe('integration: Windows Server 2022 STIG baseline', () => {
  const baseline = loadFixture();

  it('loads fixture with expected metadata', () => {
    expect(baseline.name).toBe('microsoft-windows-server-2022-stig-baseline');
    expect(baseline.title).toBe(
      'Microsoft Windows Server 2022 Security Technical Implementation Guide',
    );
    expect(baseline.requirements.length).toBe(18);
  });

  it('generates a control file for every requirement', () => {
    const profile = generateInSpecProfile(baseline);
    expect(profile.controls.size).toBe(18);
    for (const req of baseline.requirements) {
      expect(profile.controls.has(`controls/${req.id}.rb`)).toBe(true);
    }
  });

  it('generates valid Ruby control blocks', () => {
    const profile = generateInSpecProfile(baseline);
    for (const [filename, ruby] of profile.controls) {
      const id = filename.replace('controls/', '').replace('.rb', '');
      expect(ruby).toMatch(new RegExp(`^control '${id}' do\\n`));
      expect(ruby).toMatch(/\nend\n$/);
    }
  });

  it('preserves STIG metadata in control tags', () => {
    const profile = generateInSpecProfile(baseline);
    // SV-254238 has CCI, NIST, severity tags
    const sv238 = profile.controls.get('controls/SV-254238.rb')!;
    expect(sv238).toContain("tag cci: ['CCI-000366']");
    expect(sv238).toContain("tag nist: ['CM-6 b']");
    expect(sv238).toContain("tag severity: 'medium'");
    expect(sv238).toContain("tag stig_id: 'WN22-00-000010'");
  });

  it('handles controls with multiple CCIs and NIST tags', () => {
    const profile = generateInSpecProfile(baseline);
    // SV-254240 has multiple CCIs and NIST mappings
    const sv240 = profile.controls.get('controls/SV-254240.rb')!;
    expect(sv240).toContain("'CCI-000366'");
    expect(sv240).toContain("'CCI-001312'");
    expect(sv240).toContain("'CM-6 b'");
    expect(sv240).toContain("'SI-11 a'");
  });

  it('includes check and fix descriptions', () => {
    const profile = generateInSpecProfile(baseline);
    const sv238 = profile.controls.get('controls/SV-254238.rb')!;
    expect(sv238).toContain("desc 'check'");
    expect(sv238).toContain("desc 'fix'");
  });

  it('generates inspec.yml with real STIG metadata', () => {
    const profile = generateInSpecProfile(baseline);
    expect(profile.inspecYml).toContain(
      'name: microsoft-windows-server-2022-stig-baseline',
    );
    expect(profile.inspecYml).toContain(
      'title: Microsoft Windows Server 2022 Security Technical Implementation Guide',
    );
    expect(profile.inspecYml).toContain("version: '2.7.0'");
    expect(profile.inspecYml).toContain('maintainer: MITRE SAF Team');
    expect(profile.inspecYml).toContain('license: Apache-2.0');
  });

  it('generates single-file mode with all controls', () => {
    const profile = generateInSpecProfile(baseline, { singleFile: true });
    expect(profile.controls.size).toBe(1);
    const content = profile.controls.get('controls/controls.rb')!;
    for (const req of baseline.requirements) {
      expect(content).toContain(`control '${req.id}' do`);
    }
  });

  it('applies metadata overrides to real baseline', () => {
    const profile = generateInSpecProfile(baseline, {
      metadata: {
        maintainer: 'Custom Team',
        version: '99.0.0',
      },
    });
    expect(profile.inspecYml).toContain('maintainer: Custom Team');
    expect(profile.inspecYml).toContain("version: '99.0.0'");
    // Original values should be overridden
    expect(profile.inspecYml).not.toContain('maintainer: MITRE SAF Team');
    expect(profile.inspecYml).not.toContain("version: '2.7.0'");
  });

  it('handles high-impact controls correctly', () => {
    const profile = generateInSpecProfile(baseline);
    // SV-254240 has impact 0.7 (high)
    const sv240 = profile.controls.get('controls/SV-254240.rb')!;
    expect(sv240).toContain('impact 0.7');
    expect(sv240).toContain("tag severity: 'high'");
  });

  it('preserves multi-line descriptions', () => {
    const profile = generateInSpecProfile(baseline);
    // SV-254240 has a long multi-paragraph description
    const sv240 = profile.controls.get('controls/SV-254240.rb')!;
    expect(sv240).toContain('web browser');
    expect(sv240).toContain('administrative account');
  });
});
